cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,80 @@ void AddListCast(CastFunction* func) {
   DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
 }
 
+struct CastStruct {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
+    const auto out_field_count =
+        checked_cast<const StructType&>(*out->type()).num_fields();
+
+    if (in_field_count != out_field_count) {
+      return Status::TypeError("struct field sizes do not match: ",
+                               batch[0].type()->ToString(), " ", out->type()->ToString());
+    }
+
+    for (int64_t i = 0; i < in_field_count; ++i) {
+      const auto in_field_name =
+          checked_cast<const StructType&>(*batch[0].type()).field(i)->name();
+      const auto out_field_name =
+          checked_cast<const StructType&>(*out->type()).field(i)->name();
```

Comment:
Should we also check field nullability here? It seems fine to cast non-nullable to nullable, but not the other way around (unless there are no nulls).

Reply:
Done but currently failing - when casting from non-nullable to nullable do you think the null bitmap of the target should be respected or should we drop that? Right now the current test machinery expects that to be maintained from the target

Reply:
Could you expand on what you mean? Looking at the test below, why would casting a non-null value suddenly result in a null value? Changing nullability is only about whether there _can_ be nulls, not whether there actually _are_ nulls.

Reply:
Sounds good. I think I know how to get this passing

Reply:
Actually had this backwards before, but the current implementation is doing what you are describing. I think CheckCast has the wrong expectation given the below output:

```
Expected:
  -- is_valid: all not null
  -- child 0 type: int8
    [
      1,
      null
    ]
  -- child 1 type: int8
    [
      3,
      4
    ]
Actual:
  -- is_valid: all not null
  -- child 0 type: int8
    [
      1,
      2
    ]
  -- child 1 type: int8
    [
      3,
      4
    ]
```

Do I have that right? If so would you suggest trying to change CheckCast or would it be better to not use that function?

Reply:
Again, would you mind elaborating?

```
  // OK to go from non-nullable to nullable...
  std::vector<std::shared_ptr<Field>> fields1 = {
      std::make_shared<Field>("a", int8(), false),
      std::make_shared<Field>("b", int8(), false)};
  std::shared_ptr<Array> a1, b1;
  a1 = ArrayFromJSON(int8(), "[1, 2]");
  b1 = ArrayFromJSON(int8(), "[3, 4]");
  ASSERT_OK_AND_ASSIGN(auto src1, StructArray::Make({a1, b1}, fields1));

  std::vector<std::shared_ptr<Field>> fields2 = {
      std::make_shared<Field>("a", int8(), true),
      std::make_shared<Field>("b", int8(), true)};
  std::shared_ptr<Array> a2, b2;
  a2 = ArrayFromJSON(int8(), "[1, null]");
  b2 = ArrayFromJSON(int8(), "[3, 4]");
  ASSERT_OK_AND_ASSIGN(auto dest1, StructArray::Make({a2, b2}, fields2));

  CheckCast(src1, dest1);
```

I would expect the values not to change. I do not see why we are expecting a null value afterwards. We are casting `{a: int8() not null, b: int8() not null}` to {a: int8(), b: int8()}` and that should not change the _values_ at all.

Reply:
Sorry confused on my end. Had the test wrong was overlooking that the arguments to CheckCast are `input` and `expected` ; was wondering why my expected was giving me a NULL value but is obvious

Reply:
Looks good to me now.